### DISCUSSION TITLE
Add option to disable delay globally, to ensure consistency in performance of tests in jest

### DISCUSSION THREAD

**Discussion by ahayes91** [mswjs/msw#935] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**

I am finding that the performance of tests written with Mock Service Worker and @testing-library/react is not always consistent when running large numbers of tests (e.g. running `jest` locally or in a CI environment across a monorepo).

I have found that some tests will time out before the app has fully loaded (e.g. a `waitFor` query after `render` times out and shows a loading state still present in the DOM). 

Even avoiding "heavy" queries in our tests (like replacing byRole with byTestId as per https://github.com/testing-library/dom-testing-library/issues/698) show this initial loading error sometimes.

Setting up MSW in setupFilesAfterEnv has improved this performance somewhat (as per https://github.com/mswjs/examples/blob/master/examples/with-jest/jest.config.js), but I'd like to be able to further future-proof the test performance and consistency by removing the "random realistic server response time" delay MSW provides on handlers. https://mswjs.io/docs/api/context/delay

**Describe the solution you'd like**

A global configuration setting, or a setting that we can pass to `setupServer`, that allows you to disable the delay/set a default of 0 unless it has been overridden.

**Describe alternatives you've considered**

We could set `ctx.delay(0)` on every handler, but it's a lot of boilerplate as our code grows. 

**Additional context**
Some other folks mention performance with @testing-library/react and mock service worker here https://github.com/testing-library/react-testing-library/issues/819 and here https://github.com/testing-library/dom-testing-library/issues/820

**Comment by kettanaito** [comment, marked as answer]
Hey, @ahayes91. Thank you for reaching out.

First, let me clarify that there is no response delay in MSW by default. You control the time of mocked response delay explicitly using the `ctx.delay()` utility, as you've pointed out. This means that unless you are using `ctx.delay()` in your tests, there's no response delay from the library that may affect response resolution time. 

> I'd like to be able to further future-proof the test performance and consistency by removing the "random realistic server response time" delay MSW provides on handlers

Please note that this statement applies only when using `ctx.delay()` _without any arguments_. This still means you've made a conscious decision to use `ctx.delay()`, there is no implicit response delay in play. 

However, in the cases when you do use an explicit delay, it will remain the same anywhere your handler is used, including integration tests. This may be undesirable, and I generally recommend not using response delay in tests (there is no practical reason for that unless testing specific delay-dependent scenarios). 

There are numerous ways you can ensure there's no explicit delay during your test runs. **Note that these are relevant _only_ when using `ctx.delay()` in your handlers**.

## Option 1: Abstract the delay value

```js
const IS_NODE_PROCESS = determineInAnyWayYouSeeFit()

function ms(duration) {
  return IS_NODE_PROCESS ? 0 : duration
}

rest.get('/user', (req, res, ctx) => {
  // Use the response delay during development
  // but not during testing.
  return res(ctx.delay(ms(250))
})
```

## Option 2: Write a custom `delay` response transformer

```js
import { compose, context } from 'msw'

export function delay(duration) {
  // Ensure there is no response delay in tests.
  if (IS_NODE_PROCESS) {
    return compose()
  }

  return compose(context.delay(duration))
}
```
```js
import { delay } from './delay'

rest.get('/user', (req, res, ctx) => {
  return res(delay(250))
})
```

> Learn more about [Custom response transformers](https://mswjs.io/docs/basics/response-transformer#custom-transformer).

**Comment by denchen** [comment]
Hello, can I get some clarification:

Is setting `context.delay(0)` equivalent to not using `context.delay()` at all? Or are there any functional differences between the two?

**Reply by vossmalte** [reply]
yes, as stated in the [documentation](https://mswjs.io/docs/api/context/delay#implicit-response-delay), no argument uses a random delay.
